File: octave/phase.m
clear;

SF = 48000; % sample rate
F = 2; % frequency
T = 1; % time

t_in = 0:SF * T;

left = sin(t_in * 2 * pi * F / SF);
right = cos(t_in * 2 * pi * F / SF);

old_atan2 = 0;
wrap = 0;

for k = 1 + t_in

    new_atan2 = atan2(left(k), right(k));

    wrapped(k) = new_atan2;

    if old_atan2 - new_atan2 > pi
        wrap += 1
    end

    old_atan2 = new_atan2;
    unwrapped(k) = wrap * 2 * pi + new_atan2;
end

figure(1)
clf
hold on
plot(left)
plot(right)
grid on

figure(2)
clf
hold on
plot(wrapped)
plot(unwrapped)
grid on

figure(3)
clf
hold on
plot(wrapped)
grid on

File: examples/phase/phase.rs
// phase

use std::f64::consts::PI;

fn main() {
    const SF: usize = 48000; // sample rate
    const F: f64 = 2.0; // frequency
    const T: usize = 1; // time

    let mut left = vec![];
    let mut right = vec![];
    let mut wrapped = vec![];
    let mut unwrapped = vec![];

    let mut old_atan2 = 0.0;
    let mut wrap = 0.0;

    for k in 0..SF * T {
        left.push((k as f64 * 2.0 * PI * F / SF as f64).sin());
        right.push((k as f64 * 2.0 * PI * F / SF as f64).cos());

        let new_atan2 = libm::atan2(left[k], right[k]);
        wrapped.push(new_atan2);
        if old_atan2 - new_atan2 > PI {
            wrap += 1.0;
        }
        old_atan2 = new_atan2;
        unwrapped.push(wrap * 2.0 * PI + new_atan2);
    }

    phase::plot_all(left, right, wrapped, unwrapped.clone(), SF, "phase");

    phase::quality(unwrapped, "octave/phase.csv");
}
